src/v/archival/ntp_archiver_service.cc

```diff
@@ -374,6 +377,31 @@ ss::future<cloud_storage::upload_result> ntp_archiver::upload_manifest() {
     co_return co_await _remote.upload_manifest(_bucket, manifest(), fib);
 }
 
+cloud_storage::remote::reset_input_stream
+ntp_archiver::make_reset_func(upload_candidate candidate) {
+    switch (candidate.upload_kind) {
+    case upload_kind::single:
+        return [this, candidate]()
+                 -> ss::future<std::unique_ptr<storage::stream_provider>> {
+            auto handle = co_await candidate.source->reader().data_stream(
+              candidate.file_offset, candidate.final_file_offset, _io_priority);
+            co_return std::make_unique<storage::segment_reader_handle>(
+              std::move(handle));
+        };
+        break;
+    case upload_kind::multiple:
+        return [this, candidate]()
+                 -> ss::future<std::unique_ptr<storage::stream_provider>> {
+            co_return std::make_unique<storage::concat_segment_reader_view>(
```

Comment:
concat_segment_read_view does handle single segments, we also have tests to that effect in `src/v/storage/tests/concat_segment_reader_test.cc`. The only issue with using it for both cases is that the upload candidate has `sources` for the multiple file variant and `source` for the single file variant, making that decision is required to create a handle.

Although now thinking about it, it makes more sense to just keep a vector of sources in the upload candidate, which will hold zero, one or more segments and we can remove the `upload_candidate.source` entirely.

The checks like `upload.source == nullptr` can be changed to `upload.sources.empty()` and so on. This will remove the need for storing an upload kind in the candidate.

Reply:
updated the upload candidate to always have a segments vector, and always use `concat_segment_read_view`, this removes the `upload_kind` and the switches in most places in code.